### PR TITLE
test(tracer): streamline e2e test structure

### LINKED ISSUE #3028: Maintenance: streamline tracer e2e tests
### Summary

As discussed during the review of #3012 ([here](https://github.com/aws-powertools/powertools-lambda-typescript/pull/3012#pullrequestreview-2280592869)), the end to end tests for the Tracer utility have become unwieldy and have a lot of overlap. This has been the result of years of adding test cases to cover new features.

We should streamline the test structure so that the overlap & duplication is reduced.

### Why is this needed?

So that the tests are more maintainable and also so that we can reduce the number of traces emitted for each test, which in case of multiple subsequent test runs lead to excessive noise.

### Which area does this relate to?

Tests

### Solution

_No response_

### Acknowledgment

- [X] This request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

### PR DESCRIPTION
**Pull request by dreamorosi**
### Changes

> Please provide a summary of what's being changed

This PR simplifies the end to end test structure for the Tracer integration tests by reducing the amount of duplication and overlap between test cases.

The new structure, while still with some room for improvement, tries to address the main code paths & features for each instrumentation method and selectively uses some shared features in each one of them.

In addition to the changes above, I also simplified how values are passed from the test to the AWS Lambda functions ran as part of the tests. Before the change, the values were shared like this:

```mermaid
flowchart TD
    A[constants.ts] -->|imports values| B[test file]
    B -->|deploys w/ env variables| C[lambda function]
    C -->|reads env vars| D[function uses values]
    B -->|asserts that values as expected| D
```

after the change:

```mermaid
flowchart TD
    A[constants.ts] -->|imports values| B[test file]
    A -->|imports values| C[lambda function uses values]
    B -->|asserts that values as expected| C
```

While the change doesn't appear significant in its diagram representation, importing the values directly in both the tests and the function allows us to remove a considerable amount of boilerplate code needed to pass and parse environment variables on both sides.

The PR also improves the X-Ray trace utilities in `packages/testing` to retry loading the trace data when certain subsegments are not present when the trace is first loaded.

> Please add the issue number below, if no issue is present the PR might get blocked and not be reviewed

**Issue number:** closes #3028

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by dreamorosi** [comment]
I've run the [integration tests ~10 times](https://github.com/aws-powertools/powertools-lambda-typescript/actions/workflows/run-e2e-tests.yml) over the past few hours and the flakiness on the Tracer tests is greatly reduced, as well as the runtime for these tests.

There have been some transient failures in other areas (Logger & Idempotency) which consistently disappear on the next retry. I'll work on these at some point in the future but for the sake of keeping this PR focused, I'm working only on Tracer.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hi @dreamorosi! Awesome work refactoring these tests and testing what's important, I mean: handler, standalone function, SDK tracer, HTTP tracer, coldStart, annotation and others.

My only question about the tests in general is the use of `for` and logic inside these loops. I don't know if you can break it down into more small tests to test specific things like "first invocation with coldstart", "second invocation" and so on. You already have the data, so if possible, you should be more specific and reduce the cognitive load to understand the test.

Marking this PR as "Request Changes" to hear from you if we can change anything or if we merge this PR as is.

File: packages/tracer/tests/e2e/decorator.test.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    if (!process.env.DISABLE_TEARDOWN) {
+      await testStack.destroy();
+    }
+  }, TEARDOWN_TIMEOUT);
+
+  it(
+    'should generate all trace data correctly',
+    async () => {
+      // Assess
+      for (let i = 0; i < invocationCount; i++) {
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/tracer/tests/e2e/decorator.test.ts:102]: Good point - splitting it into different test cases improves maintainability

**Comment by dreamorosi** [comment]
Hey @leandrodamascena, thank you for the review & the useful suggestion.

I've broken down the tests and ran them again (green): https://github.com/aws-powertools/powertools-lambda-typescript/actions/runs/10775415657

**Review by leandrodamascena — APPROVED** [review]
APPROVED!

**Pull request merged**